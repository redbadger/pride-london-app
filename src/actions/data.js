// @flow
import type { Dispatch } from "redux";
import { getCmsData, updateCmsData } from "../integrations/cms";
import type { SavedData } from "../integrations/cms";

export type DataAction =
  | { type: "REQUEST_CMS_DATA" }
  | {
      type: "RECEIVE_CMS_DATA",
      data: SavedData
    }
  | { type: "REQUEST_UPDATE_CMS_DATA" };

export const getData = (
  hideSplashScreen: ?Function,
  getCmsDataFn: typeof getCmsData = getCmsData
) => async (dispatch: Dispatch<DataAction>) => {
  dispatch({
    type: "REQUEST_CMS_DATA"
  });

  const cmsData = await getCmsDataFn();

  dispatch({
    type: "RECEIVE_CMS_DATA",
    data: cmsData
  });

  // This feels quite dirty.
  // Once we have a proper splash screen, we can coordinate the
  // event loading and splash screen hiding better.
  if (hideSplashScreen) {
    hideSplashScreen();
  }
};

export const backgroundRefreshData = (
  updateCmsDataFn: typeof updateCmsData = updateCmsData
) => async (dispatch: Dispatch<DataAction>) => {
  const cmsData = await updateCmsDataFn();

  if (cmsData.updated) {
    // We can change this to visually notify
    // the user of new content and give them
    // the choice to explicitly load it, rather than
    // refreshing content under their hands.
    // This would be a new action type
    dispatch({
      type: "RECEIVE_CMS_DATA",
      data: cmsData
    });
  }
};

export const updateData = (
  updateCmsDataFn: typeof updateCmsData = updateCmsData
) => async (dispatch: Dispatch<DataAction>) => {
  dispatch({
    type: "REQUEST_UPDATE_CMS_DATA"
  });

  const cmsData = await updateCmsDataFn();

  dispatch({
    type: "RECEIVE_CMS_DATA",
    data: cmsData
  });
};
