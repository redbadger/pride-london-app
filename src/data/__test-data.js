// @flow
// eslint-disable-next-line import/no-extraneous-dependencies
import { gen, sampleOne as sample } from "@rgbboy/testcheck";
import type { ValueGenerator } from "@rgbboy/testcheck";
import { DateTime } from "luxon";
import { FORMAT_CONTENTFUL_ISO } from "../lib/date";
import type { Event } from "./event";
import type { FieldRef } from "./field-ref";
import type { HeaderBanner } from "./header-banner";
import type { Performance } from "./performance";
import type { Sponsor } from "./sponsor";

export const sampleOne = <A>(generator: ValueGenerator<A>): A =>
  sample(generator, 30, 1);

export const sampleArrayOf = <A>(
  generator: ValueGenerator<A>
): (number => Array<A>) => (size: number) =>
  sample(gen.array(generator, { size }), 30, 1);

export const generateFieldRef: ValueGenerator<FieldRef> = gen({
  sys: gen({
    id: gen.alphaNumString
  })
});

const baseTime = 1530964800000; // July 7, 2018 12:00:00 PM GMT+00:00
const fiveMinutes = 300000;

export const generateDateString: ValueGenerator<string> = gen.int.then(int =>
  DateTime.fromMillis(baseTime + int * int * int * fiveMinutes, {
    zone: "UTC"
  }).toFormat(FORMAT_CONTENTFUL_ISO)
);

// will change this when we refactor FieldRef
export const generateCMSFieldRef: ValueGenerator<mixed> = generateFieldRef;

export const generateHeaderBanner: ValueGenerator<HeaderBanner> = gen({
  contentType: "headerBanner",
  id: gen.alphaNumString,
  locale: "en-GB",
  revision: 1,
  fields: gen({
    heading: "heading",
    headingLine2: "headingLine2",
    subHeading: "subHeading",
    heroImage: generateFieldRef,
    backgroundColour: "#ff0000"
  })
});

export const generateCMSHeaderBanner: ValueGenerator<mixed> = gen({
  sys: {
    id: gen.alphaNumString,
    contentType: {
      sys: {
        id: "headerBanner"
      }
    },
    revision: 1
  },
  fields: {
    heading: {
      "en-GB": "heading"
    },
    headingLine2: {
      "en-GB": "headingLine2"
    },
    subHeading: {
      "en-GB": "subHeading"
    },
    heroImage: {
      "en-GB": generateFieldRef
    },
    backgroundColour: {
      "en-GB": "#ff0000"
    }
  }
});

export const generateEvent: ValueGenerator<Event> = gen({
  sys: gen({
    id: gen.alphaNumString,
    contentType: {
      sys: {
        id: "event"
      }
    },
    revision: 1
  }),
  fields: gen({
    name: { "en-GB": "name" },
    eventCategories: {
      "en-GB": ["Cabaret and Variety", "Music"]
    },
    audience: { "en-GB": ["???"] },
    startTime: { "en-GB": "2018-07-07T00:00+00:00" },
    endTime: { "en-GB": "2018-07-07T03:00+00:00" },
    location: { "en-GB": { lat: 0, lon: 0 } },
    addressLine1: { "en-GB": "addressLine1" },
    addressLine2: { "en-GB": "addressLine2" },
    city: { "en-GB": "city" },
    postcode: { "en-GB": "postcode" },
    locationName: { "en-GB": "locationName" },
    eventPriceLow: { "en-GB": 0 },
    eventPriceHigh: { "en-GB": 10 },
    accessibilityOptions: {
      "en-GB": ["accessibilityOptionsA", "accessibilityOptionsB"]
    },
    eventDescription: { "en-GB": "eventDescription" },
    accessibilityDetails: { "en-GB": "accessibilityDetails" },
    email: { "en-GB": "email" },
    phone: { "en-GB": "phone" },
    ticketingUrl: { "en-GB": "ticketingUrl" },
    venueDetails: {
      "en-GB": ["venueDetailsA", "venueDetailsB"]
    },
    individualEventPicture: gen({ "en-GB": generateFieldRef }),
    eventsListPicture: gen({ "en-GB": generateFieldRef }),
    performances: { "en-GB": [] },
    recurrenceDates: { "en-GB": [] }
  })
});

export const generatePerformance: ValueGenerator<Performance> = gen({
  contentType: "performance",
  id: gen.alphaNumString,
  locale: "en-GB",
  revision: 1,
  fields: gen({
    title: "title",
    startTime: generateDateString
  })
});

export const generateCMSPerformance: ValueGenerator<mixed> = gen({
  sys: {
    id: gen.alphaNumString,
    contentType: {
      sys: {
        id: "performance"
      }
    },
    revision: 1
  },
  fields: gen({
    title: {
      "en-GB": "title"
    },
    startTime: gen({
      "en-GB": generateDateString
    })
  })
});

export const generateSponsor: ValueGenerator<Sponsor> = gen({
  contentType: "sponsor",
  id: gen.alphaNumString,
  locale: "en-GB",
  revision: 1,
  fields: gen({
    sponsorName: "sponsorName",
    sponsorLogo: generateFieldRef,
    sponsorUrl: "sponsorUrl",
    sponsorLevel: "Headline"
  })
});

export const generateCMSSponsor: ValueGenerator<mixed> = gen({
  sys: {
    id: gen.alphaNumString,
    contentType: {
      sys: {
        id: "sponsor"
      }
    },
    revision: 1
  },
  fields: {
    sponsorName: { "en-GB": "sponsorName" },
    sponsorLogo: { "en-GB": generateFieldRef },
    sponsorUrl: { "en-GB": "sponsorUrl" },
    sponsorLevel: { "en-GB": "Headline" }
  }
});
