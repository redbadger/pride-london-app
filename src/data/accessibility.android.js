// @flow

// eslint-disable-next-line import/prefer-default-export
export const checkboxAccessibilityLabel = (label: string, checked: boolean) =>
  checked ? `checked checkbox, ${label}` : `not checked checkbox, ${label}`;
