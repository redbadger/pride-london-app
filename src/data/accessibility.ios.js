// @flow

// eslint-disable-next-line import/prefer-default-export
export const checkboxAccessibilityLabel = (label: string, checked: boolean) =>
  checked ? `${label}, checkbox, selected` : `${label}, checkbox, empty`;
