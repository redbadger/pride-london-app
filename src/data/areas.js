// Simple sqaure boundary around Central London
// Adapted from https://www.london.gov.uk/file/13839/download?token=A-jJ8CEw
export default {
  northLatBoundary: "51.530327",
  southLatBoundary: "51.493764",
  westLonBoundary: "-0.152868",
  eastLonBoundary: "-0.072527"
};
