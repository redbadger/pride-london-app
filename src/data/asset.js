// @flow

export type Asset = {
  fields: {
    title: { [string]: string },
    file: {
      [string]: {
        contentType: string,
        fileName: string,
        url: string,
        details: {
          size: number,
          image: {
            height: number,
            width: number
          }
        }
      }
    }
  },
  sys: {
    id: string,
    type: "Asset",
    revision: number
  }
};
