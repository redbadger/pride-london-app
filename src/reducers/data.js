// @flow
import type { Reducer } from "redux";
import type { DataAction } from "../actions/data";
import type { CmsEntry } from "../integrations/cms";
import type { Asset } from "../data/asset";
import { expandRecurringEventsInEntries } from "../selectors/events";

export type State = {
  entries: CmsEntry[],
  assets: Asset[],
  loading: boolean,
  refreshing: boolean
};

const defaultState = {
  entries: [],
  assets: [],
  loading: true,
  refreshing: false
};

const processEntries = entries => expandRecurringEventsInEntries(entries);

const reducer: Reducer<State, DataAction> = (
  state: State = defaultState,
  action: DataAction
) => {
  switch (action.type) {
    case "REQUEST_CMS_DATA":
      return {
        ...state,
        loading: true,
        refreshing: false
      };
    case "REQUEST_UPDATE_CMS_DATA":
      return {
        ...state,
        loading: false,
        refreshing: true
      };
    case "RECEIVE_CMS_DATA":
      return {
        loading: false,
        refreshing: false,
        entries: processEntries(action.data.entries),
        assets: action.data.assets
      };
    default:
      return state;
  }
};
export default reducer;
