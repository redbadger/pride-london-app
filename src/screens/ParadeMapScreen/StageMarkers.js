// @flow
import React, { Fragment } from "react";
import { StyleSheet } from "react-native";
import { Marker } from "react-native-maps";
import { uniqWith, eqBy } from "ramda";
import type { Event } from "../../data/event";
import type { Coordinates } from "../../constants/parade-coordinates";
import stageIconActive from "../../../assets/images/stageIconActive.png";
import stageIconInactive from "../../../assets/images/stageIconInactive.png";

type Props = {
  stages: Event[],
  activeMarker: ?string,
  onMarkerPress: (stage: Event) => void,
  onMarkerSelect: (event: { nativeEvent: { coordinate: Coordinates } }) => void
};

const StageMarkers = ({
  stages,
  onMarkerPress,
  onMarkerSelect,
  activeMarker
}: Props) => {
  if (stages.length === 0) {
    return null;
  }

  const uniqueStages = uniqWith(eqBy(stage => stage.id.split("-")[0]))(stages);

  return (
    <Fragment>
      {uniqueStages.map(stage => (
        <Marker
          coordinate={{
            longitude: stage.fields.location.lon,
            latitude: stage.fields.location.lat
          }}
          key={stage.id}
          stopPropagation
          image={
            activeMarker === stage.id ? stageIconActive : stageIconInactive
          }
          onPress={() => onMarkerPress(stage)}
          onSelect={onMarkerSelect}
          style={activeMarker === stage.id ? styles.active : styles.inactive}
        />
      ))}
    </Fragment>
  );
};

const styles = StyleSheet.create({
  inactive: {
    zIndex: 1
  },
  active: {
    // React Native Maps adds a constant to the currently open callout, but if the stage marker is active we want to move it above this
    // https://github.com/react-community/react-native-maps/blob/080678b24f886c3b8104206f2f80452ee723243a/lib/ios/AirMaps/AIRMapMarker.m#L316
    zIndex: 1001
  }
});

export default StageMarkers;
