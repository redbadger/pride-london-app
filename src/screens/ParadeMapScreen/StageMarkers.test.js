// @flow
import React from "react";
import { shallow } from "enzyme";
import { Marker } from "react-native-maps";
import type { Event } from "../../data/event";
import StageMarkers from "./StageMarkers";
import { generateEvent, sampleOne } from "../../data/__test-data";

const stage = sampleOne(generateEvent, { seed: 5728 });

describe("AmenityMarkers component", () => {
  it("renders correctly", () => {
    const output = shallow(
      <StageMarkers
        stages={[stage]}
        // eslint-disable-next-line no-unused-vars
        onMarkerPress={(_: Event) => {}}
        activeMarker={null}
        onMarkerSelect={() => {}}
      />
    );
    expect(output).toMatchSnapshot();
  });

  it("renders correctly when it's the currently selected stage", () => {
    const output = shallow(
      <StageMarkers
        stages={[stage]}
        // eslint-disable-next-line no-unused-vars
        onMarkerPress={(_: Event) => {}}
        activeMarker={stage.id}
        onMarkerSelect={() => {}}
      />
    );
    expect(output).toMatchSnapshot();
  });

  it("renders nothing when no amenities provided", () => {
    const output = shallow(
      <StageMarkers
        stages={[]}
        // eslint-disable-next-line no-unused-vars
        onMarkerPress={(_: Event) => {}}
        activeMarker={null}
        onMarkerSelect={() => {}}
      />
    );
    expect(output.children().length).toBe(0);
  });

  it("renders recurring stage event once", () => {
    const recurrenceStage = {
      ...stage,
      id: `${stage.id}-recurrence-2018-07-07`
    };

    const output = shallow(
      <StageMarkers
        stages={[stage, recurrenceStage]}
        // eslint-disable-next-line no-unused-vars
        onMarkerPress={(_: Event) => {}}
        activeMarker={null}
        onMarkerSelect={() => {}}
      />
    );

    expect(output.find("MapMarker").length).toBe(1);
  });

  it("calls onMarkerPress function when pressed", () => {
    const mockHandleMarkerPress = jest.fn();
    const output = shallow(
      <StageMarkers
        stages={[stage]}
        activeMarker={null}
        onMarkerPress={mockHandleMarkerPress}
        onMarkerSelect={() => {}}
      />
    );

    output.find(Marker).simulate("press");

    expect(mockHandleMarkerPress).toHaveBeenCalled();
  });
});
