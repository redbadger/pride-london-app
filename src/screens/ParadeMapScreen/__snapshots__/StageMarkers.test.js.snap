// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmenityMarkers component renders correctly 1`] = `
<React.Fragment>
  <MapMarker
    coordinate={
      Object {
        "latitude": 0,
        "longitude": 10,
      }
    }
    image={
      Object {
        "testUri": "../../../assets/images/stageIconInactive.png",
      }
    }
    key="QF4dTqmpn9z5ItEizAZ"
    onPress={[Function]}
    onSelect={[Function]}
    stopPropagation={true}
    style={
      Object {
        "zIndex": 1,
      }
    }
  />
</React.Fragment>
`;

exports[`AmenityMarkers component renders correctly when it's the currently selected stage 1`] = `
<React.Fragment>
  <MapMarker
    coordinate={
      Object {
        "latitude": 0,
        "longitude": 10,
      }
    }
    image={
      Object {
        "testUri": "../../../assets/images/stageIconActive.png",
      }
    }
    key="QF4dTqmpn9z5ItEizAZ"
    onPress={[Function]}
    onSelect={[Function]}
    stopPropagation={true}
    style={
      Object {
        "zIndex": 1001,
      }
    }
  />
</React.Fragment>
`;
